text/000-i18n.md

```diff
@@ -0,0 +1,92 @@
+Feature Name: i18n
+- Start Date: 2015-09-24
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+This RFC proposes to add an internationalization framework to the rust compiler.
+
+#Motivations
+
+With the growth of the number of users of Rust, localizing the compiler output would be a big plus.
+
+#Proposal
+
+We make every user-facing string in the compiler into a key-value pair.
+
+For example, the following string output by the compiler:
+
+```text
+test.rs:2:4: 2:6 error: expected one of `.`, `;`, `}`, or an operator, found `::`
+```
+
+would be created via:
+
+```Rust
+format_lang!("parser.expected_operator", expected, found)
+```
+
+We would store all translations in localization files which would contain key value pairs. For example:
+
+```
+// in english file
+parser.expected_operator = "expected one of {} or an operator, found {}"
+
+// in french file
+parser.expected_operator = "un opérateur ou l'un des symboles suivants était attendu: {}, obtenu: {}"
+```
+
+If no translation is found, the english localization is used.
+
+To use a specific language, one can run rustc as follows:
+
+```Shell
+> rustc --lang fr <other arguments>
+```
+
+##Long diagnostics
+
+For error codes / long diagnostics, we keep a different file format for convenience and instead of statically baking in the string, we do a file lookup. The only difference will be that instead of calling:
+
+```Rust
+span_err!(E0123, format!("error description {}", some_substitution));
+```
+we will write:
+
+```Rust
+span_err!(diagnostics.E0123, some_substitution);
+```
+
+There will be two diagnostics keys for this, the short and long diagnostic.
+
+```
+diagnostics.E0123.short = "error description {}"
+diagnostics.E0123.long = """foo bar baz \
+quux \
+... \
+"""
+```
+
+##Storage of localizations
+
+The localization files should be stored in a folder called i10n, which is part of the rust installation folder. By default, the english files will be there, but if you put the french files there, the option `--lang fr` will work. The folder will look like this:
```

Comment:
> With lots of keys and sub-keys you might want to consider an indexed structure

I don't think we'll have too much of that.

This particular format is basically a copy of Firefox's `.properties` files for JS l10n (they use DTDs for HTML/XUL l10n).

Firefox has tons of these, but each JS file only loads the necessary ones for performance. They don't have nested keys though.

We don't need to do what Firefox does since these are for errors and warnings -- pretty cheap to do file I/O to fetch this information.